### ISSUE TITLE
Dispose via `use` in async workflows has a race condition with cancellation

### ISSUE THREAD

**Issue by polytypic** [Microsoft/visualfsharp#1436]
The current implementation of the `use` construct of the asynchronous workflow expressions has a race condition with respect to cancellation: if cancellation is requested at the right time, a disposable may be constructed, but dispose will not be called.

Consider the implementation of [`use`](https://github.com/fsharp/fsharp/blob/c5445a48d72704313e90dbba6dd9832355aafb56/src/fsharp/FSharp.Core/control.fs#L912) in terms of [`try-finally`](https://github.com/fsharp/fsharp/blob/c5445a48d72704313e90dbba6dd9832355aafb56/src/fsharp/FSharp.Core/control.fs#L866).  Due to cancellation, the implementation of `try-finally` does not translate to an implementation of `use` that guarantees disposal.

Here is a pair of examples that demonstrates the issue:

```fs
open System
open System.Threading

do let source = new CancellationTokenSource ()
   let token = source.Token
   let example = async {
     printfn "A: Here we go"
     use! h = Async.OnCancel <| fun () -> printfn "A: We got cancelled"
     use x = source.Cancel ()
             printfn "A: Creating disposable"
             {new IDisposable with
               override this.Dispose () = printfn "A: Disposed properly"}
     do printfn "A: Never getting here"
   }
   Async.Start (example, token)

do let source = new CancellationTokenSource ()
   let token = source.Token
   let example = async {
     printfn "B: Here we go"
     use! h = Async.OnCancel <| fun () -> printfn "B: We got cancelled"
     use x = printfn "B: Creating disposable"
             {new IDisposable with
               override this.Dispose () = printfn "B: Disposed properly"}
     do source.Cancel ()
     do! async { return () }
     do printfn "B: Never getting here"
   }
   Async.Start (example, token)
```

Here is sample output obtained by running the above with `F# Interactive for F# 4.0 (Open Source Edition)`:

```
B: Here we go
A: Here we go
B: Creating disposable
A: We got cancelled
B: We got cancelled
A: Creating disposable
B: Disposed properly
```

As can be seen, in both cases disposables are created, but only in case B the disposable is disposed.  Note that the example uses synchronous code to trigger the behavior.  In actual use scenarios cancellation is triggered by code on other threads and the result is that disposal will be non-deterministically cancelled.

I would expect `use` to guarantee disposal in both cases.

**Comment by dsyme** [comment]
@polytypic Thanks for reporting this issue

**Comment by eriawan** [comment]
@dsyme this issue is reproducible on my machine using VS 2015 Update 3, Windows 10.

I hope the fix will be available for F# 4.0, not just the incoming F# 4.1.

cc @cartermp

**Comment by dsyme** [comment]
@polytypic @forki @eriawan  I'd appreciate your review of the fix for this (I know it's been merged, but I'd still appreciate it)
